[quant][graphmode][fx] Quantize the output of statically quantized fp16 op in QuantizeHandler

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58455 [quant][fx][graphmode][refactor] Remove qconfig_map from Quantizer
* #58454 [quant][graphmode][fx] Move qat_swap_modules outside of Quantizer
* #58453 [quant][fx][graphmode][refactor] Factor out generate_qconfig_map to qconfig_utils.py
* **#58445 [quant][graphmode][fx] Quantize the output of statically quantized fp16 op in QuantizeHandler**

Summary:
Previously the output of statically quantized fp16 operator is not quantized in QuantizeHandler, which is not consistent with
the behavior of static int8 operators. Also it does not work well with reference functions, this PR
changes the fp16 static QuantizeHandler to quantize (call to(torch.float16)) in the QuantizeHandler, this also
makes the future support for reference functions easier.

There is no change to the output of quantization flow except for a few cases (see changed test).

1. One case is test_reshape_fp16, previous graph is missing a dequantize() after the reshape
2. Fixed test_linear_static_fp16 since it has a missing dequantize before as well for CopyNode

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28495830](https://our.internmc.facebook.com/intern/diff/D28495830)

jerryzh168 committed:
[quant][graphmode][fx] Quantize the output of statically quantized fp16 op in QuantizeHandler

Summary:
Previously the output of statically quantized fp16 operator is not quantized in QuantizeHandler, which is not consistent with
the behavior of static int8 operators. Also it does not work well with reference functions, this PR
changes the fp16 static QuantizeHandler to quantize (call to(torch.float16)) in the QuantizeHandler, this also
makes the future support for reference functions easier.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28495830).

vkuzo commented:
> Previously the output of statically quantized fp16 operator is not quantized in QuantizeHandler, which is not consistent with
the behavior of static int8 operators.

sorry, I think it's hard to visualize this...any chance we can include an ASCII graph of the "before" and "after", for both easier review and easier time of future people who read this PR?

jerryzh168 committed:
Update on "[quant][graphmode][fx] Quantize the output of statically quantized fp16 op in QuantizeHandler"


Summary:
Previously the output of statically quantized fp16 operator is not quantized in QuantizeHandler, which is not consistent with
the behavior of static int8 operators. Also it does not work well with reference functions, this PR
changes the fp16 static QuantizeHandler to quantize (call to(torch.float16)) in the QuantizeHandler, this also
makes the future support for reference functions easier.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28495830](https://our.internmc.facebook.com/intern/diff/D28495830)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28495830).

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -3960,23 +3957,26 @@ def forward(self, x):
         expected_occurrence = {
             ns.call_function(torch.quantize_per_tensor): 2,
-            ns.call_method("dequantize"): 2,
+            # dequantize after first linear, before reshape and before output
+            ns.call_method("dequantize"): 3,
```
hmm, sorry, so is the graph actually changing (since these tests are updated with different invocation counts)?

jerryzh168 replied:
Oh, right, I updated the summary to talk about these two test cases, these are incorrect before actually, this PR fixes them.

vkuzo reviewed (approved):
makes sense.  Would it be possible to actually add to the PR summary the ascii graphs of the "before" and "after" of what is the behavior change?  It's hard to understand this from just reading the changed test cases and running the code live, and it would help future people who look at this code.

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -1302,9 +1302,8 @@ def is_output_quantized(node: Node, obj: QuantizeHandler) -> bool:
-            # converter
-            if not activation_is_int8_quantized(qconfig) or \
+            # or the activation is not statically quantized
+            if not activation_is_statically_quantized(qconfig) or \
```
@vkuzo the main change made in the PR is described in the summary already, but what's causing the change in test case is this line. previously we are saying if qconfig is not int8 we'll set quantized to False (since fp16 QuantizeHandler will output float32 Tensor), but in reality that is not correct. In the reshape_fp16 test case, we have:
```
dequantize = linear.dequantize();  linear = None
to = dequantize.to(torch.float16);  dequantize = None
reshape = to.reshape(-1, 4);  to = None
```
that is, the output of reshape is float16 (quantized to float16) instead of non-quantized.

after this PR the QuantizeHandler for reshape is producing a quantized Tensor as well (same as int8 QuantizeHandler), so we can change this condition to cover all quantized cases.
